wsdot component adopts wsdot package

ucodery opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The core API logic of the `wsdot` component has been moved to an async PyPI package. This change requires that package and removes the need for explicit request handling.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #140814
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

Note that this is a new package created for HA. It is pretty alpha, but at v0.0.0 supports exactly what was done within the entity, and has tests.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ucodery committed:
wsdot component adopts wsdot package

ucodery committed:
update generated files

joostlek reviewed (changes requested):
Awesome, let's make sure it has a higher version than 0.0.0

ucodery committed:
format code

joostlek commented on tests/components/wsdot/test_sensor.py:
```diff
@@ -23,42 +20,50 @@
 }
 
 
+@pytest.fixture
+def test_travel_time() -> wsdot.TravelTime:
+    """Return TravelTime data based on test fixture payload."""
+    test_data = load_fixture("wsdot/wsdot.json")
+    test_response = json.loads(test_data)
+    return wsdot.TravelTime(**test_response)
```
Instead, create a `conftest.py` where we `patch("homeassistant.components.wsdot.sensor.WsdotTravelTimes")` and then set the functions to return what we expect

ucodery commented on homeassistant/components/wsdot/sensor.py:
```diff
@@ -106,57 +91,36 @@ class WashingtonStateTravelTimeSensor(WashingtonStateTransportSensor):
+        if self._wsdot_travel is None:
+            session = async_get_clientsession(self.hass)
+            self._wsdot_travel = wsdot.WsdotTravelTimes(api_key=self._access_code, session=session)
+        return self._wsdot_travel
```
When would I construct the `WsdotTravelTimes`? It can't be done in `setup_platform` as it need a session, and there is no running async loop at the point the platform is created.

joostlek replied:
uhh, there definetly is a loop running there, you might need to change the signature to `async def async_setup_platform`

ucodery replied:
yeah, that worked to be able to grab the session in `async_setup_platform`

ucodery commented on tests/components/wsdot/test_sensor.py:
```diff
@@ -23,42 +20,50 @@
-    assert sensor.extra_state_attributes[ATTR_TIME_UPDATED] == datetime(
+    assert sensor.extra_state_attributes["TimeUpdated"] == datetime(
         2017, 1, 21, 15, 10, tzinfo=timezone(timedelta(hours=-8))
     )
```
I tried to use `async_setup_component`, but awaiting that doesn't create any entities: the integration's `async_setup_platform` is never called. (which is also showing that the earlier test is just asserting that nothing happening isn't raising any errors)

joostlek replied:
I can try to give this a shot

ucodery committed:
move wsdot to async_setup_platform

joostlek committed:
Fix tests

ucodery commented:
> Why is _wsdot_travel.get_travel_time typed to get an intbut we give it a str?

Ah, that is an oversight. The `self._travel_time_id` was originally a `str` and I didn't update it. Now I'm not sure if this is actually a breaking change to the configuration. In the documentation for this integration this field is specified as `id string Required`, but the example yaml indicates that maybe everyone has been using ints `- id: 95`. (https://www.home-assistant.io/integrations/wsdot#configuration)

So probably, the docs needs updated to specify `int`, and setup needs to cast whatever to `int` to cover historic cases.

joostlek commented:
I guess we need to `int` them.

I now updated the tests, can we release a version that is not 0.0.0?

ucodery committed:
cast wsdot travel id

ucodery committed:
bump wsdot to 0.0.1

ucodery commented:
I believe I have addressed all feedback. There is still a test failure in `tests/components/group/test_notify.py::test_send_message_with_data`, which seems unrelated to this PR.

joostlek commented:
Unrelated, let's put it ready for review and merge it :D
